File: src/main/java/io/myosotisdev/utopianism/util/IntegerTemplate.java
package me.rainbowmeowcat.befall.mathmatic;

import java.math.BigDecimal;
import java.util.HashMap;
import java.util.Map;

public class IntegerTemplate implements INumberTemplate<Integer>
{
    private int base;
    private final boolean useProportion;
    private final boolean useValues;
    private Map<String, ValueDouble> proportion;
    private Map<String, ValueInt> values;

    private int latest;

    public IntegerTemplate(int base)
    {
        this(base, true, true);
    }

    public IntegerTemplate(int base, boolean useProportion, boolean useValues)
    {
        this.base = base;
        this.useProportion = useProportion;
        this.useValues = useValues;
        this.proportion = new HashMap<>();
        this.values = new HashMap<>();
    }

    @Override
    public void addProportion(String key, double percentage, int duration)
    {

    }

    @Override
    public void addValue(String key, int value, int duration)
    {

    }

    @Override
    public Integer requestValue()
    {
        latest = base
                + (useProportion ? new BigDecimal(base * (proportion.values().stream().mapToDouble(value -> value.doubleValue()).reduce(0, Double::sum))).intValue() : 0)
                + (useValues ? values.values().stream().mapToInt(value -> value.intValue()).reduce(0, Integer::sum) : 0);
        return getValue();
    }

    @Override
    public Integer getValue()
    {
        return latest;
    }
}

File: src/main/kotlin/io/myosotisdev/utopianism/util/IntegerTemplate.kt
package io.myosotisdev.utopianism.util

import java.math.BigDecimal
import java.util.concurrent.ConcurrentHashMap

class IntegerTemplate @JvmOverloads constructor(private val base: Int, private val useProportion: Boolean = true, private val useValues: Boolean = true) : INumberTemplate<Int>
{
    private val proportion: ConcurrentHashMap<String, ValueDouble> = ConcurrentHashMap()
    private val values: ConcurrentHashMap<String, ValueInt> = ConcurrentHashMap()
    private val updated = false
    private var latest = 0

    override fun addProportion(key: String?, percentage: Double, duration: Int)
    {
    }

    override fun addValue(key: String?, value: Int, duration: Int)
    {
        TODO("Not yet implemented")
    }

    override fun removeProportion(key: String?)
    {
        TODO("Not yet implemented")
    }

    override fun removeValue(key: String?)
    {
        TODO("Not yet implemented")
    }

    override fun value(): Int
    {
        return if (updated) (base
                             + (if (useProportion) BigDecimal(base * proportion.values.stream()
                .mapToDouble { obj: ValueDouble -> obj.value }
                .reduce(0.0) { a: Double, b: Double -> java.lang.Double.sum(a, b) }).toInt()
        else 0)
                             + if (useValues) values.values.stream()
                .mapToInt { obj: ValueInt -> obj.value }
                .reduce(0) { a: Int, b: Int -> Integer.sum(a, b) }
                             else 0).also { latest = it }
        else latest
    }
}